### PR TITLE
Metana Bootcamp - Week 2

### PR DESCRIPTION
**Pull request by hepnerthomas**
# Metana Bootcamp: Week 2

1. Build an NFT contract and integrate it on OpenSea:
    - [x] Create an NFT that can be minted for free and has a collection of 10 items with traits and pictures.
    - [x] It should work on OpenSea per the tutorial.
    - [x] Use Goerli for gas savings (rinkeby is deprecated).
    - [x] Make sure I can mint your NFT from etherscan!

    Deployed Contract:
      - [Etherscan](https://goerli.etherscan.io/address/0x2850efc442a6a072b1aac27fa2090154b898b6ab#writeContract)
      - [OpenSea](https://testnets.opensea.io/collection/thomasnft-1)

2. Create an NFT that can be minted with an ERC20 token:
    - [x] You should create an NFT smart contract that will mint if the user pays 10 ERC20 tokens from a separate ERC20 token that you create. (use 18 decimal places as usual).
    - [x] Hint: the user will need to approve the transfer of ERC20 tokens first.

    Deployed Contract:
      - [ERC20 Contract](https://goerli.etherscan.io/address/0x8dcfa70bfd933b7e4efa2364014d2f008252bf3d)
      - [Etherscan](https://goerli.etherscan.io/address/0x95572330f8d84ae214f9203882dab4f40ef1fd88)
      - [OpenSea](https://testnets.opensea.io/collection/thomas-purchaseable-nft)

3. Staking NFTs:
    - [x] Create a contract where users can send their NFTs and withdraw 10 ERC20 tokens every 24 hours.
    - [x] The user can withdraw the NFT at any time.
    - [x] The smart contract must take possession of the NFT and only the user should be able to withdraw it. Beware of the corner case of re-staking to bypass the timer.
    - [x] IMPORTANT: you must be able to transfer the NFT like in the youtube video I recorded above

    Deployed Contract:
      - [Etherscan]()
      - [OpenSea]()

### CODE REVIEW

File: week2/remix-tutorials/NFTStaking.sol [hunk 1]
```diff
@@ -0,0 +1,38 @@
+// SPDX-License-Identifier: GPL-3.0
```
Comment (dhruvinparikh) [inline week2/remix-tutorials/NFTStaking.sol:1]: Will not review

File: week2/remix-tutorials/SimpleNFT.sol [hunk 2]
```diff
@@ -0,0 +1,66 @@
+// SPDX-License-Identifier: UNLICENSED
```
Comment (dhruvinparikh) [inline week2/remix-tutorials/SimpleNFT.sol:1]: Will not review

File: week2/remix-tutorials/simple-nft.sol [hunk 3]
```diff
@@ -0,0 +1,37 @@
+// SPDX-License-Identifier: GPL-3.0
```
Comment (dhruvinparikh) [inline week2/remix-tutorials/simple-nft.sol:1]: will not review

File: week2/nft-project/contracts/PurchaseNFT.sol [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+        IERC20 token = IERC20(TOKEN_ADDRESS);
+        uint256 senderBalance = token.balanceOf(msg.sender);
+        require(senderBalance >= TOKEN_PRICE, "Insufficient token balance.");
+        token.transferFrom(msg.sender, address(this), TOKEN_PRICE);
+
+        uint256 tokenId = totalSupply();
+        _safeMint(msg.sender, tokenId);
+    }
+
```
Comment (dhruvinparikh) [inline week2/nft-project/contracts/PurchaseNFT.sol:26]: This contract receives ERC20 token. Please add a withdraw token function only callable by owner.

File: week2/nft-project/contracts/StakeNFTForTokens.sol [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        itemNFT = _address;
+        token = new Token();
+    }
+
+    function onERC721Received(address operator, address from, uint256 tokenId, bytes calldata data) external returns (bytes4) {
+        originalOwner[tokenId] = from;
+        depositTime[tokenId] = block.timestamp;
+        lastClaimedTime[tokenId] = block.timestamp;
+        return IERC721Receiver.onERC721Received.selector;
+    }
```
Comment (dhruvinparikh) [inline week2/nft-project/contracts/StakeNFTForTokens.sol:69]:
```suggestion
    function onERC721Received(address operator, address from, uint256 tokenId, bytes calldata data) external returns (bytes4) {
        require(msg.sender == itemNFT, "unaccepted NFT collection");
        originalOwner[tokenId] = from;
        depositTime[tokenId] = block.timestamp;
        lastClaimedTime[tokenId] = block.timestamp;
        return IERC721Receiver.onERC721Received.selector;
    }
```

File: week2/nft-project/contracts/StakeNFTForTokens.sol [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+    function _claimRewards(uint256 tokenId) internal {
+        uint256 rewards = calculateRewards(tokenId);
+        if (rewards > 0) {
+            token.mintTokensToAddress(msg.sender, rewards);
+        }
+        lastClaimedTime[tokenId] = block.timestamp;
+    }
```
Comment (dhruvinparikh) [inline week2/nft-project/contracts/StakeNFTForTokens.sol:77]:
```suggestion
    function _claimRewards(uint256 tokenId) internal {
        uint256 rewards = calculateRewards(tokenId);
        if (rewards > 0) {
            token.mintTokensToAddress(msg.sender, rewards);
            lastClaimedTime[tokenId] = block.timestamp;
        }
    }
```

File: week2/nft-project/contracts/StakeNFTForTokens.sol [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        require(originalOwner[tokenId] == msg.sender, "only the owner can claim token rewards");
+        _claimRewards(tokenId);
+    }
+
+    function withdrawNFT(uint256 tokenId) external {
+        require(originalOwner[tokenId] == msg.sender, "only the owner can withdraw");
+        _claimRewards(tokenId);
+        delete depositTime[tokenId];
+        itemNFT.safeTransferFrom(address(this), msg.sender, tokenId);
+    }
```
Comment (dhruvinparikh) [inline week2/nft-project/contracts/StakeNFTForTokens.sol:89]:
```suggestion
    function withdrawNFT(uint256 tokenId) external {
        require(originalOwner[tokenId] == msg.sender, "only the owner can withdraw");
        _claimRewards(tokenId);
        delete depositTime[tokenId];
        delete originalOwner[tokenId];
        itemNFT.safeTransferFrom(address(this), msg.sender, tokenId);
    }
```

File: week2/nft-project/contracts/PurchaseNFT.sol (hunk 4 above)
Comment (hepnerthomas) [reply week2/nft-project/contracts/PurchaseNFT.sol:26]: done

**Pull request merged**